File: src/main/java/me/testaccount666/serversystem/commands/executables/time/TabCompleterTime.java
package me.testaccount666.serversystem.commands.executables.time;

import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.command.Command;

import java.util.List;
import java.util.Optional;

public class TabCompleterTime implements ServerSystemTabCompleter {

    @Override
    public Optional<List<String>> tabComplete(User commandSender, Command command, String label, String... arguments) {
        if (!PermissionManager.hasCommandPermission(commandSender, "Time.Use", false)) return Optional.of(List.of());

        if (command.getName().equalsIgnoreCase("time")) {
            if (arguments.length <= 1) {
                var possibleCompletions = List.of("day", "night", "noon", "midnight");
                var completions = possibleCompletions.stream().filter(completion -> completion.toLowerCase().startsWith(arguments[0].toLowerCase())).toList();

                return Optional.of(completions);
            }

            if (arguments.length == 2) return handleWorldCompletions(commandSender, 1, arguments);

            return Optional.of(List.of());
        }

        if (arguments.length == 1) return handleWorldCompletions(commandSender, 0, arguments);

        return Optional.of(List.of());
    }

    private Optional<List<String>> handleWorldCompletions(User commandSender, int index, String... arguments) {
        if (!PermissionManager.hasCommandPermission(commandSender, "Time.World", false)) return Optional.of(List.of());

        var possibleCompletions = Bukkit.getWorlds().stream().map(World::getName).toList();
        var completions = possibleCompletions.stream().filter(completion -> completion.toLowerCase().startsWith(arguments[index].toLowerCase())).toList();

        return Optional.of(completions);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/time/TabCompleterTime.kt
package me.testaccount666.serversystem.commands.executables.time

import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter
import me.testaccount666.serversystem.managers.PermissionManager.hasCommandPermission
import me.testaccount666.serversystem.userdata.User
import org.bukkit.Bukkit
import org.bukkit.command.Command

class TabCompleterTime : ServerSystemTabCompleter {
    override fun tabComplete(commandSender: User, command: Command, label: String, vararg arguments: String): List<String> {
        if (!hasCommandPermission(commandSender, "Time.Use", false)) return listOf()

        if (command.name.equals("time", true)) {
            if (arguments.size <= 1) {
                val possibleCompletions = listOf("day", "night", "noon", "midnight")
                return possibleCompletions.filter { it.startsWith(arguments[0], true) }
            }

            if (arguments.size == 2) return handleWorldCompletions(commandSender, 1, *arguments)
            return listOf()
        }

        if (arguments.size == 1) return handleWorldCompletions(commandSender, 0, *arguments)
        return listOf()
    }

    private fun handleWorldCompletions(commandSender: User, index: Int, vararg arguments: String): List<String> {
        if (!hasCommandPermission(commandSender, "Time.World", false)) return listOf()

        val possibleCompletions = Bukkit.getWorlds().map { it.name }
        return possibleCompletions.filter { it.startsWith(arguments[index], true) }
    }
}
